File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    int startColor = 0;
    int newColor = 0;
    
    void fill(vector<vector<int>>& img, int r, int c) {
        if (r == img.size() || r == -1 || c == img[0].size() || c == -1 || img[r][c] == newColor) // out of bounds
            return;
        
        if (img[r][c] == startColor) {
            img[r][c] = newColor;
            fill(img, r+1, c);
            fill(img, r-1, c);
            fill(img, r, c+1);
            fill(img, r, c-1);
        }
    }
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        startColor = image[sr][sc];
        newColor = color;
        
        fill(image, sr, sc);
        
        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def dfs(self, image, r, c, color):
        if r == len(image) or r == -1 or c == len(image[0]) or c == -1 or image[r][c] != self.startColor:
            return
        
        image[r][c] = color
        self.dfs(image, r+1, c, color)
        self.dfs(image, r-1, c, color)
        self.dfs(image, r, c+1, color)
        self.dfs(image, r, c-1, color)
    

    
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        if image[sr][sc] == color:
            return image
        
        self.startColor = image[sr][sc]
        self.dfs(image, sr, sc, color)
        
        return image